File: performance/tuple_operations.lua
local function slice(pyTbl, pyStart, pyStop)
  local result = {}
  local startIndex = (pyStart or 0) + 1
  if startIndex < 1 then
    startIndex = 1
  end
  local endIndex = pyStop and math.min(pyStop, #pyTbl) or #pyTbl

  for i = startIndex, endIndex do
    result[#result + 1] = pyTbl[i]
  end
  return result
end

local function concatTables(a, b)
  local result = {}
  for _, value in ipairs(a) do
    result[#result + 1] = value
  end
  for _, value in ipairs(b) do
    result[#result + 1] = value
  end
  return result
end

local iterations = 10000

-- Test 1: Tuple creation
local sum1 = 0
for _ = 1, iterations do
  local t = {1, 2, 3, 4, 5}
  sum1 = sum1 + t[1]
end
print(sum1)

-- Test 2: Tuple access
local base = {10, 20, 30, 40, 50, 60, 70, 80, 90, 100}
local sum2 = 0
for _ = 1, iterations do
  local a = base[1]
  local b = base[4]
  local c = base[8]
  local d = base[10]
  sum2 = sum2 + a + b + c + d
end
print(sum2)

-- Test 3: Tuple slicing
local source = {1, 2, 3, 4, 5, 6, 7, 8, 9, 10}
local sum3 = 0
for _ = 1, iterations do
  local s1 = slice(source, 2, 5)
  local s2 = slice(source, 0, 3)
  local s3 = slice(source, 7, nil)
  local a = s1[1]
  local b = s2[1]
  local c = s3[1]
  sum3 = sum3 + a + b + c
end
print(sum3)

-- Test 4: Tuple concatenation
local left = {1, 2, 3, 4, 5}
local right = {6, 7, 8, 9, 10}
local sum4 = 0
for _ = 1, iterations do
  local combined = concatTables(left, right)
  local a = combined[1]
  local b = combined[10]
  sum4 = sum4 + a + b
end
print(sum4)

-- Test 5: Complex operations (mixed)
local t1 = {1, 2, 3}
local t2 = {4, 5, 6}
local sum5 = 0
for _ = 1, iterations do
  local joined = concatTables(t1, t2)
  local slice_result = slice(joined, 1, 5)
  local a = slice_result[1]
  local b = slice_result[4]
  sum5 = sum5 + a + b
end
print(sum5)

-- Final checksum
local total = sum1 + sum2 + sum3 + sum4 + sum5
print("Total checksum: " .. total)

File: performance/tuple_operations.py
#!/usr/bin/env python3
"""
Performance test for tuple operations in Python
Tests: creation, access, slicing, and concatenation
"""

def main():
    iterations = 10000

    # Test 1: Tuple creation
    sum1 = 0
    for i in range(iterations):
        t = (1, 2, 3, 4, 5)
        sum1 = sum1 + t[0]
    print(sum1)

    # Test 2: Tuple access
    base = (10, 20, 30, 40, 50, 60, 70, 80, 90, 100)
    sum2 = 0
    for i in range(iterations):
        a = base[0]
        b = base[3]
        c = base[7]
        d = base[9]
        sum2 = sum2 + a + b + c + d
    print(sum2)

    # Test 3: Tuple slicing
    source = (1, 2, 3, 4, 5, 6, 7, 8, 9, 10)
    sum3 = 0
    for i in range(iterations):
        s1 = source[2:5]
        s2 = source[:3]
        s3 = source[7:]
        a = s1[0]
        b = s2[0]
        c = s3[0]
        sum3 = sum3 + a + b + c
    print(sum3)

    # Test 4: Tuple concatenation
    left = (1, 2, 3, 4, 5)
    right = (6, 7, 8, 9, 10)
    sum4 = 0
    for i in range(iterations):
        combined = left + right
        a = combined[0]
        b = combined[9]
        sum4 = sum4 + a + b
    print(sum4)

    # Test 5: Complex operations (mixed)
    t1 = (1, 2, 3)
    t2 = (4, 5, 6)
    sum5 = 0
    for i in range(iterations):
        joined = t1 + t2
        slice_result = joined[1:5]
        a = slice_result[0]
        b = slice_result[3]
        sum5 = sum5 + a + b
    print(sum5)

    # Final checksum to prevent optimization
    total = sum1 + sum2 + sum3 + sum4 + sum5
    print(f"Total checksum: {total}")


if __name__ == "__main__":
    main()
